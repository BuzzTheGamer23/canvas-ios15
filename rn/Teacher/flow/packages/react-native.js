// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
