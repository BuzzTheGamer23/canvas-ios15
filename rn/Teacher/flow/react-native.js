//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
