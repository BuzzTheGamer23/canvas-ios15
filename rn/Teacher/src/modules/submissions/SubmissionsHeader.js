//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import {
  View,
  StyleSheet,
  ActionSheetIOS,
  AlertIOS,
} from 'react-native'

import i18n from 'format-message'
import { LinkButton } from '../../common/buttons'
import { Heading1, Text } from '../../common/text'
import colors from '../../common/colors'

type SubmissionFilterOptionType = 'all' | 'late' | 'notsubmitted' | 'notgraded' | 'graded' | 'lessthan' | 'morethan' | 'cancel'

export type SubmissionFilterOption = {
  type: SubmissionFilterOptionType,
  title: string,
  filterFunc?: Function,
}

export type SelectedSubmissionFilter = {
  filter: SubmissionFilterOption,
  metadata?: ?any,
}

export type SubmissionsHeaderProps = {
  filterOptions: SubmissionFilterOption[],
  selectedFilter?: ?SelectedSubmissionFilter,
  onClearFilter: Function,
  onSelectFilter: Function,
}

const anonymousSubtitle = i18n('Anonymous grading')
const mutedSubtitle = i18n('Grades muted')
const bothSubtitle = i18n('Grades muted, Anonymous grading')

export default class SubmissionsHeader extends Component<any, SubmissionsHeaderProps, any> {

  chooseFilter = () => {
    const options = this.props.filterOptions.map((option) => option.title)
    ActionSheetIOS.showActionSheetWithOptions({
      options,
      cancelButtonIndex: options.length - 1,
      title: i18n('Filter by:'),
    }, this.updateFilter)
  }

  updateFilter = (index: number) => {
    const prompt = (title: string, callback: Function) => {
      let message = i18n('Out of {count}', { count: this.props.pointsPossible || 0 })
      AlertIOS.prompt(
        title,
        message,
        callback,
        'plain-text',
        '',
        'numeric'
      )
    }

    const filter = this.props.filterOptions[index]
    const selectedFilter: SelectedSubmissionFilter = {
      filter: filter,
    }

    switch (filter.type) {
      case 'lessthan':
      case 'morethan':
        prompt(filter.title, (text) => {
          selectedFilter.metadata = text
          this.props.onSelectFilter(selectedFilter)
        })
        return
      case 'cancel':
        break
      default:
        this.props.onSelectFilter(selectedFilter)
        break
    }
  }

  clearFilter = () => {
    this.props.onClearFilter()
  }

  render () {
    let title = i18n('All Submissions')
    const selected = this.props.selectedFilter
    if (selected && selected.filter && selected.filter.type !== 'all') {
      title = selected.filter.title
    }

    let subTitle = ''
    if (this.props.muted && this.props.anonymous) {
      subTitle = bothSubtitle
    } else if (this.props.muted) {
      subTitle = mutedSubtitle
    } else if (this.props.anonymous) {
      subTitle = anonymousSubtitle
    }

    return (<View style={styles.headerWrapper}>
              <View style={styles.header}>
                <Heading1
                  style={styles.headerTitle}
                  >
                  { title }
                </Heading1>
                { this.renderFilterButton() }
              </View>
              {!!subTitle && <Text style={styles.subtitle}>{subTitle}</Text>}
            </View>)
  }

  renderFilterButton = () => {
    let title = i18n('Filter')
    let accessibilityLabel = i18n('Filter Submissions')
    let onPress = this.chooseFilter
    const selected = this.props.selectedFilter

    if (selected &&
        selected.filter &&
        selected.filter.type !== 'all' &&
        selected.filter.type !== 'cancel') {
      title = i18n('Clear Filter')
      accessibilityLabel = i18n('Clear Filter Submissions')
      onPress = this.clearFilter
    }

    return (<LinkButton
              testID='submission-list.filter'
              onPress={onPress}
              style={styles.filterButton}
              accessibilityLabel={ accessibilityLabel }
              >
              { title }
            </LinkButton>)
  }

  // This assumes using some flavor of SubmissionRowDataProps to back the rows in a submission list
  static defaultFilterOptions (): SubmissionFilterOption[] {
    return [
      {
        type: 'all',
        title: i18n('All submissions'),
      },
      {
        type: 'late',
        title: i18n('Submitted late'),
        filterFunc: (submissions: any) => submissions.filter((s) => s.status === 'late'),
      },
      {
        type: 'notsubmitted',
        title: i18n("Haven't submitted yet"),
        filterFunc: (submissions: any) => submissions.filter((s) => s.grade === 'not_submitted'),
      },
      {
        type: 'notgraded',
        title: i18n("Haven't been graded"),
        filterFunc: (submissions: any) => submissions.filter((s) => s.grade === 'ungraded'),
      },
      {
        type: 'graded',
        title: i18n('Graded'),
        filterFunc: (submissions: any) => submissions.filter((s) => s.score !== null && s.score !== undefined),
      },
      {
        type: 'lessthan',
        title: i18n('Scored less than…'),
        filterFunc: (submissions: any, metadata: any) => submissions.filter((s) => {
          return (s.score !== null && s.score !== undefined) && (s.score < Number(metadata))
        }),
      },
      {
        type: 'morethan',
        title: i18n('Scored more than…'),
        filterFunc: (submissions: any, metadata: any) => submissions.filter((s) => {
          return (s.score !== null && s.score !== undefined) && (s.score > Number(metadata))
        }),
      },
      {
        type: 'cancel',
        title: i18n('Cancel'),
      },
    ]
  }
}

const styles = StyleSheet.create({
  headerWrapper: {
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: 'lightgrey',
    paddingTop: 16,
    paddingBottom: 8,
    paddingHorizontal: 16,
  },
  header: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    justifyContent: 'space-between',
  },
  headerTitle: {
    fontSize: 20,
    fontWeight: '600',
    color: '#2d3b44',
  },
  subtitle: {
    color: colors.grey4,
    fontSize: 14,
    fontWeight: '500',
  },
  filterButton: {
    marginBottom: 1,
  },
})
