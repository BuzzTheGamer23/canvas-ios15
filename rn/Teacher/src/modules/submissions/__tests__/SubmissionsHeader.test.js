//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow
import { ActionSheetIOS, AlertIOS } from 'react-native'
import React from 'react'
import SubmissionsHeader from '../SubmissionsHeader.js'
import renderer from 'react-test-renderer'

jest.mock('ActionSheetIOS', () => ({
  showActionSheetWithOptions: jest.fn(),
}))
jest.mock('AlertIOS', () => ({
  prompt: jest.fn(),
}))

const filterOptions = [
  {
    type: 'all',
    title: 'all',
  },
  {
    type: 'late',
    title: 'late',
    filterFunc: (submissions: any) => submissions.filter((s) => s.status === 'late'),
  },
  {
    type: 'morethan',
    title: 'morethan',
  },
  {
    type: 'lessthan',
    title: 'lessthan',
  },
  {
    type: 'cancel',
    title: 'cancel',
  }]

test('SubmissionsHeader chose filter function', () => {
  const instance = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} />
  ).getInstance()
  instance.chooseFilter()

  expect(ActionSheetIOS.showActionSheetWithOptions).toHaveBeenCalledWith({
    options: [
      'all',
      'late',
      'morethan',
      'lessthan',
      'cancel',
    ],
    cancelButtonIndex: 4,
    title: 'Filter by:',
  }, instance.updateFilter)
})

test('SubmissionsHeader update filter', () => {
  const onSelectFilter = jest.fn()
  const instance = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} onSelectFilter={onSelectFilter} />
  ).getInstance()

  instance.updateFilter(1)

  expect(onSelectFilter).toHaveBeenCalledWith({
    filter: instance.props.filterOptions[1],
  })
})

test('SubmissionsHeader update filter', () => {
  const onSelectFilter = jest.fn()
  const instance = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} onSelectFilter={onSelectFilter} />
  ).getInstance()

  let callback
  const prompt = jest.fn((title, message, cb) => {
    callback = cb
  })
  // $FlowFixMe
  AlertIOS.prompt = prompt
  instance.updateFilter(3)
  expect(prompt).toHaveBeenCalled()
  if (callback) {
    callback()
  }
  expect(onSelectFilter).toHaveBeenCalledWith({
    filter: instance.props.filterOptions[3],
  })
})

test('SubmissionsHeader cancel and clear', () => {
  const onSelectFilter = jest.fn()
  const onClearFilter = jest.fn()
  const instance = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} onSelectFilter={onSelectFilter} onClearFilter={onClearFilter} />
  ).getInstance()

  instance.updateFilter(4)
  expect(onSelectFilter).not.toHaveBeenCalled()

  instance.clearFilter()
  expect(onClearFilter).toHaveBeenCalled()
})

test('SubmissionHeader anonymous grading', () => {
  const tree = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} anonymous />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('SubmissionHeader muted grading', () => {
  const tree = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} muted />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

test('SubmissionHeader anonymous and muted', () => {
  const tree = renderer.create(
    <SubmissionsHeader filterOptions={filterOptions} anonymous muted />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})
