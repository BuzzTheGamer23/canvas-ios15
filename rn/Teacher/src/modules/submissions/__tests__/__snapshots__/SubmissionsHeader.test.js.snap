// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionHeader anonymous and muted 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "lightgrey",
      "borderBottomWidth": 0.5,
      "paddingBottom": 8,
      "paddingHorizontal": 16,
      "paddingTop": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessibilityTraits="header"
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": ".SFUIDisplay",
          },
          Object {
            "color": "#2D3B45",
            "fontFamily": ".SFUIDisplay-semibold",
            "fontSize": 20,
          },
          Object {
            "color": "#2d3b44",
            "fontSize": 20,
            "fontWeight": "600",
          },
        ]
      }
    >
      All Submissions
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel="Filter Submissions"
      accessibilityTraits="button"
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": 1,
          "opacity": 1,
        }
      }
      testID="submission-list.filter"
      tvParallaxProperties={undefined}
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "color": "#374A59",
                  "fontSize": 14,
                },
                Object {
                  "fontFamily": ".SFUIDisplay-bold",
                },
                Object {
                  "color": "#374A59",
                },
                Object {},
              ],
              Object {
                "fontFamily": ".SFUIDisplay",
              },
            ]
          }
        >
          Filter
        </Text>
      </View>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "fontFamily": ".SFUIDisplay",
        },
        Object {
          "color": "#2D3B45",
          "fontSize": 16,
        },
        Object {
          "color": "#8B969E",
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {
          "fontFamily": ".SFUIDisplay-medium",
        },
      ]
    }
  >
    Grades muted, Anonymous grading
  </Text>
</View>
`;

exports[`SubmissionHeader anonymous grading 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "lightgrey",
      "borderBottomWidth": 0.5,
      "paddingBottom": 8,
      "paddingHorizontal": 16,
      "paddingTop": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessibilityTraits="header"
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": ".SFUIDisplay",
          },
          Object {
            "color": "#2D3B45",
            "fontFamily": ".SFUIDisplay-semibold",
            "fontSize": 20,
          },
          Object {
            "color": "#2d3b44",
            "fontSize": 20,
            "fontWeight": "600",
          },
        ]
      }
    >
      All Submissions
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel="Filter Submissions"
      accessibilityTraits="button"
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": 1,
          "opacity": 1,
        }
      }
      testID="submission-list.filter"
      tvParallaxProperties={undefined}
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "color": "#374A59",
                  "fontSize": 14,
                },
                Object {
                  "fontFamily": ".SFUIDisplay-bold",
                },
                Object {
                  "color": "#374A59",
                },
                Object {},
              ],
              Object {
                "fontFamily": ".SFUIDisplay",
              },
            ]
          }
        >
          Filter
        </Text>
      </View>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "fontFamily": ".SFUIDisplay",
        },
        Object {
          "color": "#2D3B45",
          "fontSize": 16,
        },
        Object {
          "color": "#8B969E",
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {
          "fontFamily": ".SFUIDisplay-medium",
        },
      ]
    }
  >
    Anonymous grading
  </Text>
</View>
`;

exports[`SubmissionHeader muted grading 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "lightgrey",
      "borderBottomWidth": 0.5,
      "paddingBottom": 8,
      "paddingHorizontal": 16,
      "paddingTop": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessibilityTraits="header"
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": ".SFUIDisplay",
          },
          Object {
            "color": "#2D3B45",
            "fontFamily": ".SFUIDisplay-semibold",
            "fontSize": 20,
          },
          Object {
            "color": "#2d3b44",
            "fontSize": 20,
            "fontWeight": "600",
          },
        ]
      }
    >
      All Submissions
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel="Filter Submissions"
      accessibilityTraits="button"
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": 1,
          "opacity": 1,
        }
      }
      testID="submission-list.filter"
      tvParallaxProperties={undefined}
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "color": "#374A59",
                  "fontSize": 14,
                },
                Object {
                  "fontFamily": ".SFUIDisplay-bold",
                },
                Object {
                  "color": "#374A59",
                },
                Object {},
              ],
              Object {
                "fontFamily": ".SFUIDisplay",
              },
            ]
          }
        >
          Filter
        </Text>
      </View>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "fontFamily": ".SFUIDisplay",
        },
        Object {
          "color": "#2D3B45",
          "fontSize": 16,
        },
        Object {
          "color": "#8B969E",
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {
          "fontFamily": ".SFUIDisplay-medium",
        },
      ]
    }
  >
    Grades muted
  </Text>
</View>
`;
